### PR TITLE
Doc: Major Cori Update

### PR DESCRIPTION
**Pull request by ax3l**
Update Cori instructions after the major upgrade yesterday.
Before using the new modules, remove the old, pre-build software:
```bash
rm -rf $HOME/sw/knl $HOME/sw/haswell $HOME/sw/gpu $HOME/sw/cori_gpu
```

### CODE REVIEW

File: Docs/source/install/hpc/cori.rst [hunk 1]
```diff
@@ -45,25 +45,25 @@ And install ADIOS2, BLAS++ and LAPACK++:
...
-   cmake -S src/c-blosc -B src/c-blosc-knl-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/c-blosc-1.12.1-knl-install
+   cmake -S src/c-blosc -B src/c-blosc-knl-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/c-blosc-1.12.1-install
    cmake --build src/c-blosc-knl-build --target install --parallel 16
 
    # ADIOS2
    git clone -b v2.7.1 https://github.com/ornladios/ADIOS2.git src/adios2
    rm -rf src/adios2-knl-build
-   cmake -S src/adios2 -B src/adios2-knl-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/adios2-2.7.1-knl-install
+   cmake -S src/adios2 -B src/adios2-knl-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/adios2-2.7.1-install
    cmake --build src/adios2-knl-build --target install --parallel 16
```
Comment (RTSandberg) [inline Docs/source/install/hpc/cori.rst:55]: I get errors:
[ 28%] Building CXX object source/adios2/CMakeFiles/adios2_core.dir/engine/dataman/DataManMonitor.cpp.o
In file included from /global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:17:
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h:34:20: error: 'std::shared_ptr' has not been declared
   34 |     void Send(std::shared_ptr<std::vector<char>> buffer);
      |                    ^~~~~~~~~~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h:34:30: error: expected ',' or ...' before '<' token
   34 |     void Send(std::shared_ptr<std::vector<char>> buffer);
      |                              ^
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h:35:10: error: 'shared_ptr' in namespace 'std' does not name a template type
   35 |     std::shared_ptr<std::vector<char>> Receive();
      |          ^~~~~~~~~~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h:17:1: note: 'std::shared_ptr' is defined in header '<memory>'; did you forget to '#include <memory>'?
   16 | #include <thread>
  +++ |+#include <memory>
   17 | 
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:85:6: error: variable or field 'Send' declared void
   85 | void ZmqPubSub::Send(std::shared_ptr<std::vector<char>> buffer)
      |      ^~~~~~~~~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:85:27: error: 'shared_ptr' is not a member of 'std'
   85 | void ZmqPubSub::Send(std::shared_ptr<std::vector<char>> buffer)
      |                           ^~~~~~~~~~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:18:1: note: 'std::shared_ptr' is defined in header '<memory>'; did you forget to '#include <memory>'?
   17 | #include "ZmqPubSub.h"
  +++ |+#include <memory>
   18 | 
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:85:54: error: expected primary-expression before '>' token
   85 | void ZmqPubSub::Send(std::shared_ptr<std::vector<char>> buffer)
      |                                                      ^~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:85:57: error: 'buffer' was not declared in this scope; did you mean 'setbuffer'?
   85 | void ZmqPubSub::Send(std::shared_ptr<std::vector<char>> buffer)
      |                                                         ^~~~~~
      |                                                         setbuffer
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:93:6: error: 'shared_ptr' in namespace 'std' does not name a template type
   93 | std::shared_ptr<std::vector<char>> ZmqPubSub::Receive()
      |      ^~~~~~~~~~
/global/homes/r/rsand/src/adios2/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp:93:1: note: 'std::shared_ptr' is defined in header '<memory>'; did you forget to '#include <memory>'?
   93 | std::shared_ptr<std::vector<char>> ZmqPubSub::Receive()
      | ^~~
[ 28%] Building CXX object source/adios2/CMakeFiles/adios2_core.dir/engine/dataman/DataManReader.cpp.o
gmake[2]: *** [source/adios2/CMakeFiles/adios2_core.dir/build.make:1084: source/adios2/CMakeFiles/adios2_core.dir/toolkit/zmq/zmqpubsub/ZmqPubSub.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:3506: source/adios2/CMakeFiles/adios2_core.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2

File: Docs/source/install/hpc/cori.rst [hunk 2]
```diff
@@ -45,25 +45,25 @@ And install ADIOS2, BLAS++ and LAPACK++:
...
 
    # BLAS++ (for PSATD+RZ)
    git clone https://bitbucket.org/icl/blaspp.git src/blaspp
    rm -rf src/blaspp-knl-build
-   cmake -S src/blaspp -B src/blaspp-knl-build -Duse_openmp=ON -Duse_cmake_find_blas=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/blaspp-master-knl-install
+   cmake -S src/blaspp -B src/blaspp-knl-build -Duse_openmp=ON -Duse_cmake_find_blas=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/blaspp-master-install
```
Comment (RTSandberg) [inline Docs/source/install/hpc/cori.rst:60]: Don't know if this matters
-- Checking BLAS float return type
   BLAS (sdot) returns float as float (standard)
-- Checking for CBLAS library 
-- Checking for CBLAS library -lcblas
   CBLAS library not found. Tester cannot be built.
-- Checking for LAPACK library 
-- Checking for LAPACK library -llapack
   LAPACK library not found. Tester cannot be built.
CMake Warning at test/CMakeLists.txt:8 (message):
  CBLAS not found; tester cannot be built.

File: Docs/source/install/hpc/cori.rst (hunk 1 above)
Comment (ax3l) [reply Docs/source/install/hpc/cori.rst:55]: Thanks! Good catch!

Fun, missing `#include <memory>` in 2.7.1:
https://github.com/ornladios/ADIOS2/blob/v2.7.1/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h

Seems fixed for the next release:
https://github.com/ornladios/ADIOS2/issues/2709
https://github.com/ornladios/ADIOS2/blob/master/source/adios2/toolkit/zmq/zmqpubsub/ZmqPubSub.h

Can you patch manually this line? I will add this to docs later.

Comment (ax3l) [reply Docs/source/install/hpc/cori.rst:55]: Alternatively, we can add `-DADIOS2_USE_ZeroMQ=OFF` to the configure line one above. Can you check that not patching but disabling ZeroMQ also works?

File: Docs/source/install/hpc/cori.rst (hunk 2 above)
Comment (ax3l) [reply Docs/source/install/hpc/cori.rst:60]: Ye, that's ok we don't need to run the tests of BLAS++ here.

File: Docs/source/install/hpc/cori.rst [hunk 3]
```diff
@@ -45,25 +45,25 @@ And install ADIOS2, BLAS++ and LAPACK++:
...
    rm -rf src/c-blosc-knl-build
-   cmake -S src/c-blosc -B src/c-blosc-knl-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/c-blosc-1.12.1-knl-install
+   cmake -S src/c-blosc -B src/c-blosc-knl-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/c-blosc-1.12.1-install
    cmake --build src/c-blosc-knl-build --target install --parallel 16
 
    # ADIOS2
    git clone -b v2.7.1 https://github.com/ornladios/ADIOS2.git src/adios2
    rm -rf src/adios2-knl-build
-   cmake -S src/adios2 -B src/adios2-knl-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/adios2-2.7.1-knl-install
+   cmake -S src/adios2 -B src/adios2-knl-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/adios2-2.7.1-install
```
Comment (ax3l) [inline Docs/source/install/hpc/cori.rst]: Due to https://github.com/ornladios/ADIOS2/issues/2709 in the 2.7.1 release:
```suggestion
   cmake -S src/adios2 -B src/adios2-knl-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/adios2-2.7.1-install
```

File: Docs/source/install/hpc/cori.rst [hunk 4]
```diff
@@ -173,12 +173,12 @@ And install ADIOS2:
...
    git clone -b v1.21.1 https://github.com/Blosc/c-blosc.git src/c-blosc
    rm -rf src/c-blosc-gpu-build
-   cmake -S src/c-blosc -B src/c-blosc-gpu-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/c-blosc-1.12.1-gpu-install
+   cmake -S src/c-blosc -B src/c-blosc-gpu-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/cori_gpu/c-blosc-1.12.1-install
    cmake --build src/c-blosc-gpu-build --target install --parallel 16
 
    git clone -b v2.7.1 https://github.com/ornladios/ADIOS2.git src/adios2
    rm -rf src/adios2-gpu-build
-   cmake -S src/adios2 -B src/adios2-gpu-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/adios2-2.7.1-gpu-install
+   cmake -S src/adios2 -B src/adios2-gpu-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/cori_gpu/adios2-2.7.1-install
```
Comment (ax3l) [inline Docs/source/install/hpc/cori.rst]: Due to https://github.com/ornladios/ADIOS2/issues/2709 in the 2.7.1 release:
```suggestion
   cmake -S src/adios2 -B src/adios2-gpu-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/cori_gpu/adios2-2.7.1-install
```

File: Docs/source/install/hpc/cori.rst [hunk 5]
```diff
@@ -108,13 +108,13 @@ And install ADIOS2, BLAS++ and LAPACK++:
...
    rm -rf src/c-blosc-haswell-build
-   cmake -S src/c-blosc -B src/c-blosc-haswell-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/c-blosc-1.12.1-haswell-install
+   cmake -S src/c-blosc -B src/c-blosc-haswell-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/haswell/c-blosc-1.12.1-install
    cmake --build src/c-blosc-haswell-build --target install --parallel 16
 
    # ADIOS2
    git clone -b v2.7.1 https://github.com/ornladios/ADIOS2.git src/adios2
    rm -rf src/adios2-haswell-build
-   cmake -S src/adios2 -B src/adios2-haswell-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/adios2-2.7.1-haswell-install
+   cmake -S src/adios2 -B src/adios2-haswell-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/haswell/adios2-2.7.1-install
```
Comment (ax3l) [inline Docs/source/install/hpc/cori.rst]: Due to https://github.com/ornladios/ADIOS2/issues/2709 in the 2.7.1 release:
```suggestion
   cmake -S src/adios2 -B src/adios2-haswell-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/haswell/adios2-2.7.1-install
```

File: Docs/source/install/hpc/cori.rst (hunk 2 above)
Comment (RTSandberg) [reply Docs/source/install/hpc/cori.rst:60]: Build was fine, though

File: Docs/source/install/hpc/cori.rst (hunk 1 above)
Comment (RTSandberg) [reply Docs/source/install/hpc/cori.rst:55]: Haven't checked patch yet, but `-DADIOS2_USE_ZeroMQ=OFF` in the configure flag leads to successful build

File: Docs/source/install/hpc/cori.rst [hunk 6]
```diff
@@ -45,25 +45,25 @@ And install ADIOS2, BLAS++ and LAPACK++:
...
+   cmake -S src/blaspp -B src/blaspp-knl-build -Duse_openmp=ON -Duse_cmake_find_blas=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/blaspp-master-install
    cmake --build src/blaspp-knl-build --target install --parallel 16
 
    # LAPACK++ (for PSATD+RZ)
    git clone https://bitbucket.org/icl/lapackpp.git src/lapackpp
    rm -rf src/lapackpp-knl-build
-   CXXFLAGS="-DLAPACK_FORTRAN_ADD_" cmake -S src/lapackpp -B src/lapackpp-knl-build -Duse_cmake_find_lapack=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DLAPACK_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/lapackpp-master-knl-install
+   CXXFLAGS="-DLAPACK_FORTRAN_ADD_" cmake -S src/lapackpp -B src/lapackpp-knl-build -Duse_cmake_find_lapack=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DLAPACK_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/lapackpp-master-install
```
Comment (RTSandberg) [inline Docs/source/install/hpc/cori.rst:66]: I got the following:
@cori08:~> CXXFLAGS="-DLAPACK_FORTRAN_ADD_" cmake -S src/lapackpp -B src/lapackpp-knl-build -Duse_cmake_find_lapack=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DLAPACK_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/knl/lapackpp-master-install
-- The CXX compiler identification is GNU 11.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/gcc/11.2.0/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using CMAKE_INSTALL_PREFIX = /global/homes/r/rsand/sw/knl/lapackpp-master-install
-- lapackpp_id = 1c003ef
-- Check for BLAS++
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
   Found BLAS++: /global/homes/r/rsand/sw/knl/blaspp-master-install/lib64/blaspp
-- Found BLAS: /opt/cray/pe/libsci/20.09.1/GNU/8.1/x86_64/lib/libsci_gnu.a  
-- Found LAPACK: /opt/cray/pe/libsci/20.09.1/GNU/8.1/x86_64/lib/libsci_gnu.a  
-- Checking LAPACK version
   LAPACK version 3.06.01 (30601)
-- Checking for XBLAS
   XBLAS not found.
-- Checking for LAPACKE library 
-- Checking for LAPACKE library -llapacke
   LAPACKE library not found. Tester cannot be built.
CMake Warning at test/CMakeLists.txt:8 (message):
  CBLAS not found; tester cannot be built.


-- Configuring done

Comment (RTSandberg) [reply Docs/source/install/hpc/cori.rst:66]: But the build is fine

**Review by RTSandberg — COMMENTED** [review]
I get odd build warnings:
`/global/homes/r/rsand/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp:393:17: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  393 |             *p2 = tmp;
      |             ~~~~^~~~~
`
I don't know if these are significant

Build is successful!

File: Tools/machines/cori-nersc/knl_warpx.profile.example [hunk 7]
```diff
@@ -1,20 +1,20 @@
...
-export CMAKE_PREFIX_PATH=$HOME/sw/adios2-2.7.1-knl-install:$CMAKE_PREFIX_PATH
-export CMAKE_PREFIX_PATH=$HOME/sw/blaspp-master-knl-install:$CMAKE_PREFIX_PATH
-export CMAKE_PREFIX_PATH=$HOME/sw/lapackpp-master-knl-install:$CMAKE_PREFIX_PATH
+export CMAKE_PREFIX_PATH=$HOME/sw/knl/c-blosc-1.12.1-install:$CMAKE_PREFIX_PATH
+export CMAKE_PREFIX_PATH=$HOME/sw/knl/adios2-2.7.1-install:$CMAKE_PREFIX_PATH
+export CMAKE_PREFIX_PATH=$HOME/sw/knl/blaspp-master-install:$CMAKE_PREFIX_PATH
+export CMAKE_PREFIX_PATH=$HOME/sw/knl/lapackpp-master-install:$CMAKE_PREFIX_PATH
 
-if [ -d "$HOME/sw/venvs/knl_warpx" ]
+if [ -d "$HOME/sw/haswell/venvs/knl_warpx" ]
```
Comment (RTSandberg) [inline Tools/machines/cori-nersc/knl_warpx.profile.example]: Should this be `if [ -d "$HOME/sw/knl/venvs/knl_warpx" ]` ?

File: Tools/machines/cori-nersc/knl_warpx.profile.example [hunk 8]
```diff
@@ -1,20 +1,20 @@
...
 then
-  source $HOME/sw/venvs/knl_warpx/bin/activate
+  source $HOME/sw/haswell/venvs/knl_warpx/bin/activate
```
Comment (RTSandberg) [inline Tools/machines/cori-nersc/knl_warpx.profile.example]: Similarly, `source $HOME/sw/haswell/venvs/knl_warpx/bin/activate` ?

File: Tools/machines/cori-nersc/knl_warpx.profile.example (hunk 7 above)
Comment (ax3l) [reply Tools/machines/cori-nersc/knl_warpx.profile.example]: Good catch, thank you!
```suggestion
if [ -d "$HOME/sw/knl/venvs/knl_warpx" ]
```

File: Tools/machines/cori-nersc/knl_warpx.profile.example (hunk 8 above)
Comment (ax3l) [reply Tools/machines/cori-nersc/knl_warpx.profile.example]:
```suggestion
  source $HOME/sw/knl/venvs/knl_warpx/bin/activate
```

**Comment by ax3l** [comment]
Thank you for the review @RTSandberg, all updated :)

The warning you are seeing is a buffer overflow in AMReX: https://github.com/AMReX-Codes/amrex/pull/2660

**Review by RTSandberg — APPROVED** [review]
These look good to me; I could get back up and running.

**Review by EZoni — APPROVED** [review]
Thank you for the update, Axel! And thanks @RTSandberg for reviewing the PR and for testing the build on Cori!

**Pull request merged**